File: Medium/2_AddTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
const addTwoNumbers = function (l1, l2) {
    let dummyHead = new ListNode(0, null);
    let temp = dummyHead;

    let i = l1;
    let j = l2;
    let carry = 0;

    while (i || j) {
        const x = i ? i.val : 0;
        const y = j ? j.val : 0;

        const sum = x + y + carry;

        temp.next = new ListNode(sum % 10, null);
        temp = temp.next;
        carry = Math.floor(sum / 10);

        i = i ? i.next : null;
        j = j ? j.next : null;
    }

    if (carry >= 1) {
        temp.next = new ListNode(carry, null);
    }

    return dummyHead.next;
};

File: Medium/2_add-two-numbers.java
package Medium;

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyNode = new ListNode(-1);
        ListNode head = dummyNode;

        int carry = 0;

        while (l1 != null || l2 != null) {
            int num1 = l1 != null ? l1.val : 0;
            int num2 = l2 != null ? l2.val : 0;

            int sum = num1 + num2 + carry;
            int value = sum % 10;
            carry = sum / 10;

            ListNode newNode = new ListNode(value);
            head.next = newNode;
            head = head.next;

            if (l1 != null) {
                l1 = l1.next;
            }

            if (l2 != null) {
                l2 = l2.next;
            }
        }

        if (carry != 0) {
            head.next = new ListNode(carry);
        }

        return dummyNode.next;
    }
}
